Plugins/Onnx/src/OnnxRuntimeBase.cpp

```diff
@@ -11,102 +11,126 @@
     Ort::TypeInfo inputTypeInfo = m_session->GetInputTypeInfo(i);
     auto tensorInfo = inputTypeInfo.GetTensorTypeAndShapeInfo();
     m_inputNodeDims = tensorInfo.GetShape();
-    // fix for symbolic dim = -1 from python
-    for (size_t j = 0; j < m_inputNodeDims.size(); j++) {
-      if (m_inputNodeDims[j] < 0) {
-        m_inputNodeDims[j] = 1;
-      }
-    }
   }
 
-  // get the names of the output nodes
+  // Get the names of the output nodes
   size_t numOutputNodes = m_session->GetOutputCount();
-
-  // iterate over all output nodes and get the name
+  // Iterate over all output nodes and get the name
   for (size_t i = 0; i < numOutputNodes; i++) {
     m_outputNodeNamesAllocated.push_back(
         m_session->GetOutputNameAllocated(i, allocator));
     m_outputNodeNames.push_back(m_outputNodeNamesAllocated.back().get());
 
-    // get the dimensions of the output nodes
+    // Get the dimensions of the output nodes
     // here we assume that all output nodes have the dimensions
     Ort::TypeInfo outputTypeInfo = m_session->GetOutputTypeInfo(i);
     auto tensorInfo = outputTypeInfo.GetTensorTypeAndShapeInfo();
     m_outputNodeDims = tensorInfo.GetShape();
-    // fix for symbolic dim = -1 from python
-    for (size_t j = 0; j < m_outputNodeDims.size(); j++) {
-      if (m_outputNodeDims[j] < 0) {
-        m_outputNodeDims[j] = 1;
-      }
-    }
   }
 }
 
-// inference function using ONNX runtime
-// the function assumes that the model has 1 input node and 1 output node
+// Inference function using ONNX runtime for one single entry
 std::vector<float> Acts::OnnxRuntimeBase::runONNXInference(
     std::vector<float>& inputTensorValues) const {
-  // create input tensor object from data values
+  std::vector<std::vector<float>> vectorInput;
+  vectorInput.push_back(inputTensorValues);
+  auto vectorOutput = runONNXInference(vectorInput);
+  return vectorOutput[0];
+}
+
+// Inference function using ONNX runtime
+// the function assumes that the model has 1 input node and 1 output node
+std::vector<std::vector<float>> Acts::OnnxRuntimeBase::runONNXInference(
+    std::vector<std::vector<float>>& inputTensorValues) const {
+  int batchSize = inputTensorValues.size();
+  std::vector<int64_t> inputNodeDims = m_inputNodeDims;
+  std::vector<int64_t> outputNodeDims = m_outputNodeDims;
+
+  // The first dim node should correspond to the batch size
+  // If it is -1, it is dynamic and should be set to the input size
+  if (inputNodeDims[0] == -1) {
+    inputNodeDims[0] = batchSize;
+  }
+  if (outputNodeDims[0] == -1) {
+    outputNodeDims[0] = batchSize;
+  }
+
+  if (batchSize != 1 &&
+      (inputNodeDims[0] != batchSize || outputNodeDims[0] != batchSize)) {
+    throw std::runtime_error(
+        "runONNXInference: batch size doesn't match the input or output node "
+        "size");
+  }
+
+  // Format the input as a single vector
+  std::vector<float> tensorValues;
+  for (auto track : inputTensorValues) {
```

Comment:
If we would pass the data not in form of a `std::vector<std::vector<float>>` but as an `Eigen::Array` with dynamic size NxM this would have multiple advantages:
* We could directly pass the data pointer to Onnx without copying them into another vector (if the row-major / column major thing is correct)
* We would save some memory allocations I guess
* It would also easy to check that the size of the input is correct, and it would not be so easy to pass invalid input to the function (e.g., the inner `std::vector`s could have different length).

Reply:
That seems like an elegant solution, I usually try to avoid using Eigen if not needed but I will give it a try !

Reply:
I think something like `boost::multi_array<float, 2>` would maybe be an even better solution, so you could try this as well...
But even though Eigen is not optimal, most people in our field have experience with it which is an advantage.

Reply:
Done, for your information de default of the `Eigen::Array` are CollumnMajor but change it is quite easy !